src/libsyntax/parse/parser.rs

```diff
@@ -5319,6 +5355,163 @@ impl<'a> Parser<'a> {
+        // having succeeded in parsing:
+        //
+        // *Upcoming tokens:* `::Output>;`
+        // *Unmatched count:* 3
+        // *`parse_path_segment` calls deep:* 2
+        //
+        // This will continue until the next `>` character which will also return successfully
+        // to the parent invocation of this function and decrement the count:
+        //
+        // *Upcoming tokens:* `;`
+        // *Unmatched count:* 2
+        // *`parse_path_segment` calls deep:* 1
+        //
+        // At this point, this function will expect to find another matching `>` character but
+        // won't be able to and will return an error. This will continue all the way up the
+        // call stack until the first invocation:
+        //
+        // *Upcoming tokens:* `;`
+        // *Unmatched count:* 2
+        // *`parse_path_segment` calls deep:* 0
+        //
+        // In doing this, we have managed to work out how many unmatched leading left angle
+        // brackets there are, but we cannot recover as the unmatched angle brackets have
+        // already been consumed. To remedy this, whenever `parse_generic_args` is invoked, we
+        // make a snapshot of the current parser state and invoke it on that and inspect
+        // the result:
+        //
+        // - If success (ie. when it found a matching `>` character) then the snapshot state
+        //   is kept (this is required to propagate the count upwards).
+        //
+        // - If error and in was in a recursive call, then the snapshot state is kept (this is
+        //   required to propagate the count upwards).
+        //
+        // - If error and this was the first invocation (before any recursion had taken place)
+        //   then we choose not to keep the snapshot state - that way we haven't actually
+        //   consumed any of the `<` characters, but can still inspect the count from the
+        //   snapshot to know how many `<` characters to remove. Using this information, we can
+        //   emit an error and consume the extra `<` characters before attempting to parse
+        //   the generic arguments again (this time hopefullt successfully as the unmatched `<`
+        //   characters are gone).
+        //
+        // In practice, the recursion of this function is indirect and there will be other
+        // locations that consume some `<` characters - as long as we update the count when
+        // this happens, it isn't an issue.
+        let mut snapshot = self.clone();
+        debug!("parse_generic_args_with_leading_angle_bracket_recovery: (snapshotting)");
+        match snapshot.parse_generic_args() {
+            Ok(value) => {
+                debug!(
+                    "parse_generic_args_with_leading_angle_bracket_recovery: (snapshot success) \
+                     snapshot.count={:?}",
+                    snapshot.unmatched_angle_bracket_count,
+                );
+                mem::replace(self, snapshot);
```

Comment:
I see now what you're doing. You are advancing the cloned snapshot instead of `self` and replacing depending on what has happened. I wonder if optimizing the happy path (removing the need of `mem::replace` if `parse_generic_args().is_ok()` would have a significant perf implication (for error recovery we're more lenient in how much we push our perf envelope).

Reply:
>  I wonder if optimizing the happy path (removing the need of `mem::replace` if `parse_generic_args().is_ok()`) would have a significant perf implication

I can't quite picture the optimization you are describing. One idea I've just had though (and apologies if this is what you've described) is that we could probably get away with only creating a snapshot once in the first invocation and performing all of the recursions on that.

Reply:
My point is cloning the parser into snapshot, continue using `self`. Doing that the only case where a `mem::replace` is needed is when hitting a `parse_generic_args` error. Wouldn't change much.

Reply:
I've pushed the new commit with this approach.